Container.Remove() from multiple goroutines causes panic

### Describe the bug:
Calling Container.Remove() from multiple goroutines at the exact same time causes a panic
Container.Add() seems to be working fine (unless you're trying to add 10k+ items on a 2 fps display then it crashes)
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x58 pc=0x7bd2b9]

goroutine 78 [running]:
fyne.io/fyne/v2/layout.(*boxLayout).Layout(0xc0003b19d8, {0xc0004e8000, 0x7d, 0xc000167250?}, {0x48732e?, 0x0?})
        /home/ryan/go/pkg/mod/fyne.io/fyne/v2@v2.2.2/layout/boxlayout.go:63 +0x399
fyne.io/fyne/v2.(*Container).layout(...)
        /home/ryan/go/pkg/mod/fyne.io/fyne/v2@v2.2.2/container.go:186
fyne.io/fyne/v2.(*Container).Remove(0xc00037fe40, {0xb0fe98?, 0xc0003f9e00})
        /home/ryan/go/pkg/mod/fyne.io/fyne/v2@v2.2.2/container.go:139 +0x1fb
main.removeLater({0xb0fe98, 0xc0003f9e00})
        /home/ryan/Downloads/FyneContainerCrash/main.go:43 +0x54
created by main.main
        /home/ryan/Downloads/FyneContainerCrash/main.go:29 +0xf0
exit status 2
```

### To Reproduce:
Example code here: https://go.dev/play/p/pSAsuB7XU-5

Steps to reproduce the behaviour:
1. Create a container
2. Add a ton of objects to that container
3. Launch a goroutine for each object
4. Have each goroutine remove an object from the container at the same time
5. Panic

### Device (please complete the following information):
 - **OS:** Linux (amd64)
 - **Version:** 5.18.7-arch1-1
 - **Go version:** 1.18.3
 - **Fyne version:** 2.2.1

--- a/container.go
+++ b/container.go
@@ -1,5 +1,9 @@
 package fyne
 
+import (
+	"sync"
+)
+
 // Declare conformity to CanvasObject
 var _ CanvasObject = (*Container)(nil)
 
@@ -10,8 +14,9 @@ type Container struct {
 	position Position // The current position of the Container
 	Hidden   bool     // Is this Container hidden
 
-	Layout  Layout         // The Layout algorithm for arranging child CanvasObjects
-	Objects []CanvasObject // The set of CanvasObjects this container holds
+	Layout      Layout // The Layout algorithm for arranging child CanvasObjects
+	objectsLock *sync.Mutex
+	Objects     []CanvasObject // The set of CanvasObjects this container holds
 }
 
 // NewContainer returns a new Container instance holding the specified CanvasObjects.
@@ -27,7 +32,8 @@ func NewContainer(objects ...CanvasObject) *Container {
 // Deprecated: Use container.NewWithoutLayout() instead
 func NewContainerWithoutLayout(objects ...CanvasObject) *Container {
 	ret := &Container{
-		Objects: objects,
+		Objects:     objects,
+		objectsLock: &sync.Mutex{},
 	}
 
 	ret.size = ret.MinSize()
@@ -40,8 +46,9 @@ func NewContainerWithoutLayout(objects ...CanvasObject) *Container {
 // Deprecated: Use container.New() instead
 func NewContainerWithLayout(layout Layout, objects ...CanvasObject) *Container {
 	ret := &Container{
-		Objects: objects,
-		Layout:  layout,
+		Objects:     objects,
+		objectsLock: &sync.Mutex{},
+		Layout:      layout,
 	}
 
 	ret.size = layout.MinSize(objects)
@@ -126,6 +133,8 @@ func (c *Container) Remove(rem CanvasObject) {
 		return
 	}
 
+	c.objectsLock.Lock()
+	defer c.objectsLock.Unlock()
 	for i, o := range c.Objects {
 		if o != rem {
 			continue

--- a/container.go
+++ b/container.go
@@ -60,6 +60,8 @@ func (c *Container) Add(add CanvasObject) {
 		return
 	}
 
+	c.objectsLock.Lock()
+	defer c.objectsLock.Unlock()
 	c.Objects = append(c.Objects, add)
 	c.layout()
 }

--- a/container.go
+++ b/container.go
@@ -13,8 +13,8 @@ type Container struct {
 	Hidden   bool     // Is this Container hidden
 
 	Layout      Layout // The Layout algorithm for arranging child CanvasObjects
-	objectsLock sync.Mutex
-	Objects     []CanvasObject // The set of CanvasObjects this container holds
+	lock    sync.Mutex
+	Objects []CanvasObject // The set of CanvasObjects this container holds
 }
 
 // NewContainer returns a new Container instance holding the specified CanvasObjects.
@@ -60,8 +60,8 @@ func (c *Container) Add(add CanvasObject) {
 		return
 	}
 
-	c.objectsLock.Lock()
-	defer c.objectsLock.Unlock()
+	c.lock.Lock()
+	defer c.lock.Unlock()
 	c.Objects = append(c.Objects, add)
 	c.layout()
 }
@@ -131,8 +131,8 @@ func (c *Container) Remove(rem CanvasObject) {
 		return
 	}
 
-	c.objectsLock.Lock()
-	defer c.objectsLock.Unlock()
+	c.lock.Lock()
+	defer c.lock.Unlock()
 	for i, o := range c.Objects {
 		if o != rem {
 			continue

--- a/container.go
+++ b/container.go
@@ -1,8 +1,6 @@
 package fyne
 
-import (
-	"sync"
-)
+import "sync"
 
 // Declare conformity to CanvasObject
 var _ CanvasObject = (*Container)(nil)
@@ -15,7 +13,7 @@ type Container struct {
 	Hidden   bool     // Is this Container hidden
 
 	Layout      Layout // The Layout algorithm for arranging child CanvasObjects
-	objectsLock *sync.Mutex
+	objectsLock sync.Mutex
 	Objects     []CanvasObject // The set of CanvasObjects this container holds
 }
 
@@ -32,8 +30,7 @@ func NewContainer(objects ...CanvasObject) *Container {
 // Deprecated: Use container.NewWithoutLayout() instead
 func NewContainerWithoutLayout(objects ...CanvasObject) *Container {
 	ret := &Container{
-		Objects:     objects,
-		objectsLock: &sync.Mutex{},
+		Objects: objects,
 	}
 
 	ret.size = ret.MinSize()
@@ -46,9 +43,8 @@ func NewContainerWithoutLayout(objects ...CanvasObject) *Container {
 // Deprecated: Use container.New() instead
 func NewContainerWithLayout(layout Layout, objects ...CanvasObject) *Container {
 	ret := &Container{
-		Objects:     objects,
-		objectsLock: &sync.Mutex{},
-		Layout:      layout,
+		Objects: objects,
+		Layout:  layout,
 	}
 
 	ret.size = layout.MinSize(objects)

--- a/container.go
+++ b/container.go
@@ -12,7 +12,7 @@ type Container struct {
 	position Position // The current position of the Container
 	Hidden   bool     // Is this Container hidden
 
-	Layout      Layout // The Layout algorithm for arranging child CanvasObjects
+	Layout  Layout // The Layout algorithm for arranging child CanvasObjects
 	lock    sync.Mutex
 	Objects []CanvasObject // The set of CanvasObjects this container holds
 }

--- a/container_test.go
+++ b/container_test.go
@@ -128,6 +128,20 @@ func TestContainer_Remove_Race(t *testing.T) {
 	assert.Equal(t, 0, len(container.Objects))
 }
 
+func TestContainer_Add_Race(t *testing.T) {
+	container := NewContainerWithLayout(new(customLayout))
+	wg := &sync.WaitGroup{}
+	wg.Add(100)
+	for i := 0; i < 100; i++ {
+		go func() {
+			container.Add(new(dummyObject))
+			wg.Done()
+		}()
+	}
+	wg.Wait()
+	assert.Equal(t, 100, len(container.Objects))
+}
+
 func TestContainer_RemoveAll(t *testing.T) {
 	box1 := new(dummyObject)
 	box2 := new(dummyObject)

--- a/container_test.go
+++ b/container_test.go
@@ -1,6 +1,7 @@
 package fyne
 
 import (
+	"sync"
 	"testing"
 
 	"github.com/stretchr/testify/assert"
@@ -104,6 +105,27 @@ func TestContainer_Remove(t *testing.T) {
 	assert.Equal(t, float32(10), container.MinSize().Height)
 	assert.Equal(t, float32(0), box2.Position().X)
 	assert.Equal(t, float32(0), box2.Position().Y)
+}
+
+func TestContainer_Remove_Race(t *testing.T) {
+	var objs []CanvasObject
+	for i := 0; i < 100; i++ {
+		objs = append(objs, new(dummyObject))
+	}
+
+	container := NewContainerWithLayout(new(customLayout), objs...)
+
+	wg := &sync.WaitGroup{}
+	wg.Add(100)
+	for _, o := range objs {
+		rmo := o
+		go func() {
+			container.Remove(rmo)
+			wg.Done()
+		}()
+	}
+	wg.Wait()
+	assert.Equal(t, 0, len(container.Objects))
 }
 
 func TestContainer_RemoveAll(t *testing.T) {

